### PR TITLE
FLUID-5949: Implementation and test cases for promise-based fixture types in IoC testing framework

### PR DESCRIPTION
**Pull request by amb26**
New core utility fluid.firstDefined elevated from Model Transformations framework, also includes fix for FLUID-5753 allowing a testEnvironment to act as its own testCaseHolder. Improved factoring for "failing tests" infrastructure for jqUnit

### CODE REVIEW

File: tests/test-core/utils/js/IoCTestUtils.js [hunk 1]
```diff
@@ -498,8 +553,8 @@ var fluid_2_0_0 = fluid_2_0_0 || {};
             var last = pos === that.count;
             if (last) {
                 if (thisExec.execute) {
-                    fluid.test.execExecutor(thisExec, thisText);
-                    finishSequence();
+                    fluid.test.execExecutor(thisExec, thisText, finishSequence);
+                    //finishSequence();
```
Comment (simonbates) [inline tests/test-core/utils/js/IoCTestUtils.js]: Should this commented out line be removed?

File: tests/test-core/utils/js/IoCTestUtils.js [hunk 2]
```diff
@@ -519,11 +574,11 @@ var fluid_2_0_0 = fluid_2_0_0 || {};
...
                 if (nextExec.bind) {
                     that.execute(); // bind first [ODD]
-                    fluid.test.execExecutor(thisExec, thisText);
+                    fluid.test.execExecutor(thisExec, thisText, fluid.identity);
                 }
                 else {
-                    fluid.test.execExecutor(thisExec, thisText);
-                    that.execute();
+                    fluid.test.execExecutor(thisExec, thisText, that.execute);
+                    //that.execute();
```
Comment (simonbates) [inline tests/test-core/utils/js/IoCTestUtils.js]: Should this commented out line be removed?

**Comment by simonbates** [comment]
Pull request looks great. I've added some very minor comments.

**Comment by amb26** [comment]
Hi @simonbates - I've merged up with master and cleaned up. #755 is also ready for a look now - cheers!

**Comment by simonbates** [comment]
Merged at 499e6098c0151226c0989dbdc263b43635ca4521

**Pull request merged**